File: sdk/c/src/providers/mffac.c
/**
 * MFFAC — https://www.mffac.com/api
 */
#include "tempmail_internal.h"
#include <curl/curl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define MF_BASE "https://www.mffac.com/api"

static const char *mf_post_headers[] = {
    "Accept: */*",
    "Content-Type: application/json",
    "Origin: https://www.mffac.com",
    "Referer: https://www.mffac.com/",
    "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
    "sec-ch-ua: \"Chromium\";v=\"146\", \"Not-A.Brand\";v=\"24\", \"Microsoft "
    "Edge\";v=\"146\"",
    "sec-ch-ua-mobile: ?0",
    "sec-ch-ua-platform: \"Windows\"",
    "sec-fetch-dest: empty",
    "sec-fetch-mode: cors",
    "sec-fetch-site: same-origin",
    NULL};

static const char *mf_get_headers[] = {
    "Accept: */*",
    "Origin: https://www.mffac.com",
    "Referer: https://www.mffac.com/",
    "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
    "sec-ch-ua: \"Chromium\";v=\"146\", \"Not-A.Brand\";v=\"24\", \"Microsoft "
    "Edge\";v=\"146\"",
    "sec-ch-ua-mobile: ?0",
    "sec-ch-ua-platform: \"Windows\"",
    "sec-fetch-dest: empty",
    "sec-fetch-mode: cors",
    "sec-fetch-site: same-origin",
    NULL};

static char *mf_escape(const char *s) {
  if (!s)
    return tm_strdup("");
  CURL *c = curl_easy_init();
  if (!c)
    return tm_strdup(s);
  char *e = curl_easy_escape(c, s, 0);
  curl_easy_cleanup(c);
  if (!e)
    return tm_strdup(s);
  char *d = tm_strdup(e);
  curl_free(e);
  return d;
}

static void mf_local_part(const char *email, char *out, size_t cap) {
  if (!email || cap < 2) {
    if (cap)
      out[0] = '\0';
    return;
  }
  const char *at = strchr(email, '@');
  if (at && at > email) {
    size_t n = (size_t)(at - email);
    if (n >= cap)
      n = cap - 1;
    memcpy(out, email, n);
    out[n] = '\0';
  } else {
    strncpy(out, email, cap - 1);
    out[cap - 1] = '\0';
  }
}

static char *mf_received_at_to_iso(const cJSON *item) {
  if (!cJSON_IsNumber(item) || item->valuedouble <= 0)
    return tm_strdup("");
  time_t sec = (time_t)(item->valuedouble + 0.5);
  char buf[48];
  struct tm tmu;
#ifdef _WIN32
  if (gmtime_s(&tmu, &sec) != 0)
    memset(&tmu, 0, sizeof(tmu));
#else
  gmtime_r(&sec, &tmu);
#endif
  strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%SZ", &tmu);
  return tm_strdup(buf);
}

static cJSON *mf_fetch_detail(const char *id) {
  if (!id || !id[0])
    return NULL;
  char *esc = mf_escape(id);
  if (!esc)
    return NULL;

  char url[640];
  snprintf(url, sizeof(url), "%s/emails/%s", MF_BASE, esc);
  free(esc);

  tm_http_response_t *resp =
      tm_http_request(TM_HTTP_GET, url, mf_get_headers, NULL, 15);
  if (!resp || resp->status < 200 || resp->status >= 300) {
    tm_http_response_free(resp);
    return NULL;
  }

  cJSON *root = cJSON_Parse(resp->body);
  tm_http_response_free(resp);
  if (!root)
    return NULL;

  const cJSON *ok = cJSON_GetObjectItemCaseSensitive(root, "success");
  cJSON *email = cJSON_GetObjectItemCaseSensitive(root, "email");
  if (!cJSON_IsBool(ok) || !cJSON_IsTrue(ok) || !cJSON_IsObject(email)) {
    cJSON_Delete(root);
    return NULL;
  }

  cJSON *copy = cJSON_Duplicate(email, 1);
  cJSON_Delete(root);
  return copy;
}

static cJSON *mf_flatten_email(const cJSON *raw, const char *recipient) {
  cJSON *flat = cJSON_CreateObject();
  if (!flat)
    return NULL;

  cJSON *id = cJSON_GetObjectItemCaseSensitive((cJSON *)raw, "id");
  cJSON *from = cJSON_GetObjectItemCaseSensitive((cJSON *)raw, "fromAddress");
  cJSON *to = cJSON_GetObjectItemCaseSensitive((cJSON *)raw, "toAddress");
  cJSON *subject = cJSON_GetObjectItemCaseSensitive((cJSON *)raw, "subject");
  cJSON *text = cJSON_GetObjectItemCaseSensitive((cJSON *)raw, "textContent");
  cJSON *html = cJSON_GetObjectItemCaseSensitive((cJSON *)raw, "htmlContent");
  cJSON *received_at =
      cJSON_GetObjectItemCaseSensitive((cJSON *)raw, "receivedAt");
  cJSON *is_read = cJSON_GetObjectItemCaseSensitive((cJSON *)raw, "isRead");

  char *date = mf_received_at_to_iso(received_at);
  cJSON_AddStringToObject(flat, "id", TM_JSON_STR(id, ""));
  cJSON_AddStringToObject(flat, "from", TM_JSON_STR(from, ""));
  cJSON_AddStringToObject(flat, "to", TM_JSON_STR(to, recipient));
  cJSON_AddStringToObject(flat, "subject", TM_JSON_STR(subject, ""));
  cJSON_AddStringToObject(flat, "text", TM_JSON_STR(text, ""));
  cJSON_AddStringToObject(flat, "html", TM_JSON_STR(html, ""));
  cJSON_AddStringToObject(flat, "date", date ? date : "");
  cJSON_AddBoolToObject(flat, "isRead",
                        cJSON_IsBool(is_read) ? cJSON_IsTrue(is_read) : 0);
  cJSON_AddItemToObject(flat, "attachments", cJSON_CreateArray());
  free(date);
  return flat;
}

tm_email_info_t *tm_provider_mffac_generate(void) {
  const char *body = "{\"expiresInHours\":24}";
  tm_http_response_t *resp = tm_http_request(TM_HTTP_POST, MF_BASE "/mailboxes",
                                             mf_post_headers, body, 15);
  if (!resp || resp->status < 200 || resp->status >= 300) {
    tm_http_response_free(resp);
    return NULL;
  }

  cJSON *root = cJSON_Parse(resp->body);
  tm_http_response_free(resp);
  if (!root)
    return NULL;

  const cJSON *ok = cJSON_GetObjectItemCaseSensitive(root, "success");
  if (!cJSON_IsBool(ok) || !cJSON_IsTrue(ok)) {
    cJSON_Delete(root);
    return NULL;
  }

  cJSON *mb = cJSON_GetObjectItemCaseSensitive(root, "mailbox");
  if (!cJSON_IsObject(mb)) {
    cJSON_Delete(root);
    return NULL;
  }

  const char *addr =
      cJSON_GetStringValue(cJSON_GetObjectItemCaseSensitive(mb, "address"));
  const char *mid =
      cJSON_GetStringValue(cJSON_GetObjectItemCaseSensitive(mb, "id"));
  if (!addr || !addr[0] || !mid || !mid[0]) {
    cJSON_Delete(root);
    return NULL;
  }

  char email[320];
  snprintf(email, sizeof(email), "%s@mffac.com", addr);

  tm_email_info_t *info = tm_email_info_new();
  info->channel = CHANNEL_MFFAC;
  info->email = tm_strdup(email);
  info->token = tm_strdup(mid);

  const cJSON *ca = cJSON_GetObjectItemCaseSensitive(mb, "createdAt");
  if (cJSON_IsString(ca) && ca->valuestring)
    info->created_at = tm_strdup(ca->valuestring);

  cJSON_Delete(root);
  return info;
}

tm_email_t *tm_provider_mffac_get_emails(const char *token, const char *email,
                                         int *count) {
  (void)token;
  *count = -1;
  if (!email || !email[0])
    return NULL;

  char local[256];
  mf_local_part(email, local, sizeof(local));
  if (!local[0])
    return NULL;

  char *esc = mf_escape(local);
  char url[640];
  snprintf(url, sizeof(url), "%s/mailboxes/%s/emails", MF_BASE, esc);
  free(esc);

  tm_http_response_t *resp =
      tm_http_request(TM_HTTP_GET, url, mf_get_headers, NULL, 15);
  if (!resp || resp->status < 200 || resp->status >= 300) {
    tm_http_response_free(resp);
    return NULL;
  }

  cJSON *root = cJSON_Parse(resp->body);
  tm_http_response_free(resp);
  if (!root)
    return NULL;

  const cJSON *ok = cJSON_GetObjectItemCaseSensitive(root, "success");
  if (!cJSON_IsBool(ok) || !cJSON_IsTrue(ok)) {
    cJSON_Delete(root);
    return NULL;
  }

  cJSON *arr = cJSON_GetObjectItemCaseSensitive(root, "emails");
  int n = (arr && cJSON_IsArray(arr)) ? cJSON_GetArraySize(arr) : 0;
  *count = n;
  if (n == 0) {
    cJSON_Delete(root);
    return NULL;
  }

  tm_email_t *emails = tm_emails_new(n);
  if (!emails) {
    cJSON_Delete(root);
    *count = -1;
    return NULL;
  }

  for (int i = 0; i < n; i++) {
    cJSON *raw = cJSON_GetArrayItem(arr, i);
    const char *id =
        TM_JSON_STR(cJSON_GetObjectItemCaseSensitive(raw, "id"), "");
    cJSON *detail = id[0] ? mf_fetch_detail(id) : NULL;
    cJSON *flat = mf_flatten_email(detail ? detail : raw, email);
    emails[i] = tm_normalize_email(flat ? flat : raw, email);
    if (flat)
      cJSON_Delete(flat);
    if (detail)
      cJSON_Delete(detail);
  }
  cJSON_Delete(root);
  return emails;
}

File: sdk/rust/src/providers/mffac.rs
/*!
 * MFFAC — https://www.mffac.com/api
 */

use crate::config::{block_on, get_current_ua, http_client};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};
use serde_json::json;
use serde_json::Value;

const BASE: &str = "https://www.mffac.com/api";

fn mffac_headers(b: wreq::RequestBuilder, with_json_ct: bool) -> wreq::RequestBuilder {
    let mut b = b
        .header("User-Agent", get_current_ua())
        .header("Accept", "*/*")
        .header("Origin", "https://www.mffac.com")
        .header("Referer", "https://www.mffac.com/")
        .header(
            "sec-ch-ua",
            r#""Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146""#,
        )
        .header("sec-ch-ua-mobile", "?0")
        .header("sec-ch-ua-platform", r#""Windows""#)
        .header("sec-fetch-dest", "empty")
        .header("sec-fetch-mode", "cors")
        .header("sec-fetch-site", "same-origin");
    if with_json_ct {
        b = b.header("Content-Type", "application/json");
    }
    b
}

pub fn generate_email() -> Result<EmailInfo, String> {
    block_on(async {
        let resp = mffac_headers(
            http_client()
                .post(format!("{}/mailboxes", BASE))
                .body(r#"{"expiresInHours":24}"#.to_string()),
            true,
        )
        .send()
        .await
        .map_err(|e| e.to_string())?;
        if !resp.status().is_success() {
            return Err(format!("mffac generate {}", resp.status()));
        }
        let data: Value = resp.json().await.map_err(|e| e.to_string())?;
        if data.get("success").and_then(|x| x.as_bool()) != Some(true) {
            return Err("mffac: create failed".into());
        }
        let mb = data
            .get("mailbox")
            .ok_or_else(|| "mffac: no mailbox".to_string())?;
        let addr = mb
            .get("address")
            .and_then(|x| x.as_str())
            .unwrap_or("")
            .trim();
        let id = mb.get("id").and_then(|x| x.as_str()).unwrap_or("").trim();
        if addr.is_empty() || id.is_empty() {
            return Err("mffac: bad mailbox".into());
        }
        let email = format!("{}@mffac.com", addr);
        let expires_at = mb
            .get("expiresAt")
            .and_then(|v| v.as_i64().or_else(|| v.as_f64().map(|f| f as i64)));
        let created_at = mb
            .get("createdAt")
            .and_then(|x| x.as_str())
            .map(|s| s.to_string());
        Ok(EmailInfo {
            channel: Channel::Mffac,
            email,
            token: Some(id.to_string()),
            expires_at,
            created_at,
        })
    })
}

fn received_at_to_iso(raw: &Value) -> String {
    let Some(value) = raw.get("receivedAt") else {
        return String::new();
    };
    let seconds = value
        .as_i64()
        .or_else(|| value.as_f64().map(|n| n as i64))
        .or_else(|| {
            value
                .as_str()
                .and_then(|s| s.trim().parse::<f64>().ok())
                .map(|n| n as i64)
        })
        .unwrap_or(0);
    if seconds <= 0 {
        return String::new();
    }
    chrono::DateTime::from_timestamp(seconds, 0)
        .map(|dt| dt.to_rfc3339())
        .unwrap_or_default()
}

fn flatten_email(raw: &Value, recipient: &str) -> Value {
    json!({
        "id": raw.get("id").cloned().unwrap_or(Value::String(String::new())),
        "from": raw.get("fromAddress").cloned().unwrap_or(Value::String(String::new())),
        "to": raw.get("toAddress").cloned().unwrap_or(Value::String(recipient.to_string())),
        "subject": raw.get("subject").cloned().unwrap_or(Value::String(String::new())),
        "text": raw.get("textContent").cloned().unwrap_or(Value::String(String::new())),
        "html": raw.get("htmlContent").cloned().unwrap_or(Value::String(String::new())),
        "date": received_at_to_iso(raw),
        "isRead": raw.get("isRead").cloned().unwrap_or(Value::Bool(false)),
        "attachments": [],
    })
}

async fn fetch_email_detail(id: &str) -> Result<Option<Value>, String> {
    let url = format!("{}/emails/{}", BASE, urlencoding::encode(id));
    let resp = mffac_headers(http_client().get(&url), false)
        .send()
        .await
        .map_err(|e| e.to_string())?;
    if !resp.status().is_success() {
        return Ok(None);
    }
    let data: Value = resp.json().await.map_err(|e| e.to_string())?;
    if data.get("success").and_then(|x| x.as_bool()) != Some(true) {
        return Ok(None);
    }
    Ok(data.get("email").cloned().filter(|x| x.is_object()))
}

pub fn get_emails(email: &str, _token: Option<&str>) -> Result<Vec<Email>, String> {
    let local = email.split('@').next().unwrap_or(email).trim();
    let enc = urlencoding::encode(local);
    let url = format!("{}/mailboxes/{}/emails", BASE, enc);
    block_on(async {
        let resp = mffac_headers(http_client().get(&url), false)
            .send()
            .await
            .map_err(|e| e.to_string())?;
        if !resp.status().is_success() {
            return Err(format!("mffac emails {}", resp.status()));
        }
        let data: Value = resp.json().await.map_err(|e| e.to_string())?;
        if data.get("success").and_then(|x| x.as_bool()) != Some(true) {
            return Err("mffac: list failed".into());
        }
        let rows = data
            .get("emails")
            .and_then(|x| x.as_array())
            .cloned()
            .unwrap_or_default();
        let mut out = Vec::with_capacity(rows.len());
        for raw in rows {
            let mut source = raw.clone();
            if let Some(id) = raw
                .get("id")
                .and_then(|x| x.as_str())
                .map(|s| s.trim())
                .filter(|s| !s.is_empty())
            {
                if let Ok(Some(detail)) = fetch_email_detail(id).await {
                    source = detail;
                }
            }
            out.push(normalize_email(&flatten_email(&source, email), email));
        }
        Ok(out)
    })
}
